pkg/controller/certificaterequests/acme/acme.go

```diff
@@ -165,37 +165,47 @@ func (a *ACME) Sign(ctx context.Context, cr *v1.CertificateRequest, issuer v1.Ge
 		return nil, nil
 	}
 
-	// Order valid, return cert. The calling controller will update with ready if its happy with the cert.
-	if order.Status.State == cmacme.Valid {
-		x509Cert, err := pki.DecodeX509CertificateBytes(order.Status.Certificate)
-		if errors.IsInvalidData(err) {
-			log.Error(err, "failed to decode x509 certificate data on Order resource")
-			return nil, a.acmeClientV.Orders(order.Namespace).Delete(context.TODO(), order.Name, metav1.DeleteOptions{})
-		}
-		ok, err := pki.PublicKeyMatchesCertificate(csr.PublicKey, x509Cert)
-		if err != nil {
-			return nil, err
-		}
-		if !ok {
-			log.Error(err, "failed to decode x509 certificate data on Order resource, recreating...")
-			return nil, a.acmeClientV.Orders(order.Namespace).Delete(context.TODO(), order.Name, metav1.DeleteOptions{})
-		}
+	if order.Status.State != cmacme.Valid {
+		// We update here to just pending while we wait for the order to be resolved.
+		a.reporter.Pending(cr, nil, "OrderPending",
+			fmt.Sprintf("Waiting on certificate issuance from order %s/%s: %q",
+				expectedOrder.Namespace, order.Name, order.Status.State))
 
-		log.V(logf.InfoLevel).Info("certificate issued")
+		log.V(logf.DebugLevel).Info("acme Order resource is not in a ready state, waiting...")
 
-		return &issuerpkg.IssueResponse{
-			Certificate: order.Status.Certificate,
-		}, nil
+		return nil, nil
 	}
 
-	// We update here to just pending while we wait for the order to be resolved.
-	a.reporter.Pending(cr, nil, "OrderPending",
-		fmt.Sprintf("Waiting on certificate issuance from order %s/%s: %q",
-			expectedOrder.Namespace, order.Name, order.Status.State))
+	if len(order.Status.Certificate) == 0 {
+		a.reporter.Pending(cr, nil, "OrderPending",
+			fmt.Sprintf("Waiting for order-controller to add certificate data to Order %s/%s",
+				expectedOrder.Namespace, order.Name))
+
+		log.V(logf.DebugLevel).Info("Order controller has not added certificate data to the Order, waiting...")
+		return nil, nil
```

Comment:
I'm also doubtful about places where I return `nil, nil` in this PR, and in the current code.
Returning nil, nil means that the CertificateRequest *will not* be re-queued,
but nor will this controller be triggered by status changes in related ACME Orders,

If this Status.Condition is  set once, it will cause the parent controller to perform  an UpdateStatus, which in turn will trigger this function again.
But if the second time round, the `order.Status.Certificate` is still not set, then the condition will not change and the certificaterequest will not be re-queued.

So we should probably return an error here, and above.

Reply:
Ignore this comment. I had not spotted that the controller does in fact get triggered by changes to owned orders.